dts: arm: st: h7/h7rs: add memory-regions property to mac node

bklaric1 opened the pull request:
## Changes
- Added the `memory-regions` property to `stm32h7.dtsi` and `stm32h7rs` with `sram2` as default, in order to explicitly locate ethernet buffer and descriptor to an sram region inside the DT. 
- Added override for `sram2` with `sram3` where possible.
- Added `#memory-region-cells` to `sram2` and `sram3` nodes, since it is needed by the `memory-regions` property.
- Added a check for memory-regions before the check for and definition of the `sram_eth_node` in `mpu_regions.c` and `sections.ld` to both h7 and h7rs SoCs.
- Added `memory-regions` property to `dts/bindings/ethernet/st,stm32h7-ethernet.yaml`.
- Added seperate compatible for h5 series, in form of `dts/bindings/ethernet/st,stm32h5-ethernet.yaml`.

## ~~Dependencies~~
~~- #99296 (due to the addition of the `memory-regions` property to `dts/bindings/ethernet/st,stm32h7-ethernet.yaml`)~~

## Notes
@etienne-lms suggested this change in #99296.

bklaric1 committed:
dts: arm: st: h7: add memory-regions property to mac node

Use the memory-regions property in stm32h7.dtsi to explicitly locate
ethernet buffer and descriptor to an sram region. Set sram2 as default
and override with sram3 when available, in soc level .dtsi files.

Add #memory-region-cells to sram2 and sram3 nodes in specific .dtsi
files in dts/arm/st/h7, since it is needed by the memory-regions
property.

Add a check for memory-regions before the check and definition of the
sram_eth_node in mpu_regions.c and sections.ld.

Signed-off-by: Benjamin Klaric <benjamin.klaric01@gmail.com>

juickar reviewed (commented):
@bklaric1 try to rebase on main. A fix has been pushed for the CI failed test

bklaric1 commented:
@juickar they seem to be queued now for almost a day. Was there another fix in the meantime?

juickar commented:
> @juickar they seem to be queued now for almost a day. Was there another fix in the meantime?

@bklaric1 No, but from what I witnessed CI also fails to launch on other PRs

erwango reviewed (changes requested):
Thanks for the change, but in order to avoid impacting existing users, I'd prefer you keep the changes out of the boards.

So please assign the memory regions (`memory-regions = <&sram2/3>;`) in .dtsi files and keep existing pattern of using sram3 when available otherwise sram2.
This was, out of tree users won't see any difference.

etienne-lms commented on soc/st/stm32/stm32h7x/mpu_regions.c:
```diff
@@ -40,6 +36,12 @@ static const struct arm_mpu_region mpu_regions[] = {
+BUILD_ASSERT(!DT_SAME_NODE(sram_eth_node, DT_CHOSEN(zephyr_sram)),
+	     "Remove property memory-regions from 'mac' node "
+	     "to locate ethernet buffers in system RAM.");
+#endif
```
I think we should prevent one to set `memory-regions` property to the `zephyr,sram` phandle.
I'd suggest:
```patch
 #include <zephyr/devicetree.h>
 #include <zephyr/arch/arm/mpu/arm_mpu_mem_cfg.h>
 
+#if DT_NODE_HAS_STATUS_OKAY(DT_NODELABEL(mac))
+#define sram_eth_node DT_PHANDLE(DT_NODELABEL(mac), memory_regions)
+
+BUILD_ASSERT(DT_NODE_HAS_PROP(DT_NODELABEL(mac), memory_regions) &&
+	     DT_NODE_HAS_STATUS_OKAY(sram_eth_node),
+	     "Ethernet MAC node must provide an enabled memory region for ethernet buffers");
+BUILD_ASSERT(!DT_SAME_NODE(sram_eth_node, DT_CHOSEN(zephyr_sram)),
+	     "Ethernet buffer memory-region cannot be located in Zephyr system RAM.");
+#endif /* mac node enabled */
+
 static const struct arm_mpu_region mpu_regions[] = {
	(...)

 #ifdef sram_eth_node
 	MPU_REGION_ENTRY("SRAM_ETH_BUF",		
 					 DT_REG_ADDR(sram_eth_node),
 					 REGION_RAM_NOCACHE_ATTR(REGION_16K)),
 	MPU_REGION_ENTRY("SRAM_ETH_DESC",
 					 DT_REG_ADDR(sram_eth_node),
 					 REGION_PPB_ATTR(REGION_256B)),
 #endif
 };

 (...)
```

I guess `DT_NODE_HAS_PROP(DT_NODELABEL(mac), memory_regions)` assertion can be removed if the property is tagged `required`.

bklaric1 commented:
@erwango @etienne-lms I amended the commit to add the requested changes, update the property to be required, and update the commit message.
Now the `stm32h7.dtsi` defines `memory-regions` and sets it to `sram2` by default.
`stm32h7a3.dtsi` has `sram3` override for a and b series (these series don't support Ethernet as of now, but the addition of `memory-regions` to `stm32h7.dtsi` requires the addition of this).  
`stm32h743.dtsi` has `sram3` override for 43 and 53 series.  
`stm32h745.dtsi` has `sram3` override for 45, 47, 55 and 57 series.  
I hope I didn't forget anything/understood something wrong. In any case, I am open to feedback.

bklaric1 commented:
The failed twister-build (5) is related to missing `memory-regions` from `stm32h7rs.dtsi`. As originally stated, this PR depends on  #99296, but maybe the dependency should be the other way around.
Should I then add everything related to the `memory-regions` with this PR and then just use it in another PR?

bklaric1 commented on dts/arm/st/h5/stm32h562.dtsi:
```diff
@@ -66,12 +66,14 @@
 			compatible = "zephyr,memory-region", "mmio-sram";
 			reg = <0x20040000 DT_SIZE_K(64)>;
 			zephyr,memory-region = "SRAM2";
+			#memory-region-cells = <0>;
```
As far as I can tell, this is needed because memory-regions is defined as phandle-array here:
https://github.com/zephyrproject-rtos/zephyr/blob/e6fd157a93db7297583d7b93f9a74c0e0977bc95/dts/bindings/reserved-memory/memory-region.yaml#L7-L10

I don't know if it's possible to define it here directly (as <0>), and if it's wanted here, since <0> might not always be the use case.

I hope I didn't miss the point of this comment.

mathieuchopstm commented on soc/st/stm32/stm32h7x/mpu_regions.c:
```diff
@@ -21,23 +28,14 @@ static const struct arm_mpu_region mpu_regions[] = {
 					 REGION_RAM_NOCACHE_ATTR(REGION_16K)),
 	MPU_REGION_ENTRY("SRAM_ETH_DESC",
 					 DT_REG_ADDR(sram_eth_node),
 					 REGION_PPB_ATTR(REGION_256B)),
```
I assume there is a reason for which `PPB` was used rather than `RAM_NOCACHE` as done above, but I don't know which. So let's leave it as is for now...

etienne-lms commented on dts/arm/st/h7/stm32h743.dtsi:
```diff
@@ -25,12 +25,18 @@
 };
+
+&mac {
+	memory-regions = <&sram3>;
```
Names used in `memory-region-names` are rather identifiers for the phandle indices used in a `memory-regions` property. It's useful when there can be several phandles, not a single one, see for examples `clocks = ...` and `clock-names`.
Since here we only expect one, I don't think it's really useful.

bklaric1 committed:
dts: arm: st: h7rs: add memory-regions property to mac node

Add the memory-regions property to mac in stm32h7rs.dtsi,
same as in stm32h7.dtsi, and set sram2 as default. Also add
the #memory-region-cells to sram2 and sram1 nodes.

Override the sram2 as default with sram1 in stm32h7s3.dtsi and
stm32h7s3.dtsi since sram2 is disabled per default.

Add the same checks as for h7 series to mpu_regions.c and sections.c.

Signed-off-by: Benjamin Klaric <benjamin.klaric01@gmail.com>

bklaric1 committed:
dts: bindings: ethernet: add st,stm32h5-ethernet compatible

Add the compatible for h5 series for ethernet. The h5 series needs
it own because the memory-regions property isn't currently supported
by the implementation.

Add the h5 compatible to stm32h563.dtsi, in mac node.

Adjust the description in h7 compatible to remove the mention of h5
series, as now h5 series has it's own.

Signed-off-by: Benjamin Klaric <benjamin.klaric01@gmail.com>

etienne-lms commented:
> I will also add the changes for h5 series as a seperate commit and remove them from the first (also edit the commit message body), but first we need to agree on what needs to be added for h5 series.

To address H5, talking with @erwango, the simplest may be to not set `memory-regions` as `required` in **st,stm32h7-ethernet.yaml**. Since this P-R adds the property in H7 and H7RS SoC DTSI files, the default config is the one we expect (a dedicated SRAM for eth buffers on these SoCs). Board developers on H7 and H7RS willing to locate ethernet buffers in Zephyr main RAM would still need to tweak the implementation as they currently need to, at least no regression foreseen.

bklaric1 commented:
@etienne-lms I pushed a commit few minutes ago, adding seperate compatible for h5 series. Sorry, was too impatient :)
Could you check if this addition works as well or if I should just drop the `required: true` and leave it be, as you stated?

etienne-lms commented on dts/arm/st/h5/stm32h563.dtsi:
```diff
@@ -27,12 +27,13 @@
 				clocks = <&rcc STM32_CLOCK(AHB1, 20)>,
 					 <&rcc STM32_CLOCK(AHB1, 21)>;
 				status = "disabled";
+				memory-regions = <&sram2>;
```
Could you remove this property. H5 is currently expected to locate ethernet buffer in Zephyr main RAM.

etienne-lms reviewed (commented):
I see you already propose a h5 dedicate compatible. It does the jobs. Discard my previous comment and let's see what other reviews think.

bklaric1 commented on dts/arm/st/h5/stm32h563.dtsi:
```diff
@@ -27,12 +27,13 @@
 
 			mac: ethernet {
-				compatible = "st,stm32h7-ethernet", "st,stm32-ethernet";
+				compatible = "st,stm32h5-ethernet", "st,stm32-ethernet";
```
Done.

etienne-lms reviewed (commented):
To prevent build failures for example when playing with `git bisect`, could you place commit
*"dts: bindings: ethernet: add st,stm32h5-ethernet compatible"* before commit
*"dts: arm: st: h7: add memory-regions property to mac node"* that make `memory-regions` DT property required.

bklaric1 commented on dts/arm/st/h7rs/stm32h7rs.dtsi:
```diff
@@ -938,6 +940,7 @@
 				clocks = <&rcc STM32_CLOCK(AHB1, 16)>,
 					 <&rcc STM32_CLOCK(AHB1, 17)>;
 				status = "disabled";
+				memory-regions = <&sram2>;
```
Done.
Also reordered the commits, so that the "*dts: bindings: ethernet: add st,stm32h5-ethernet compatible*" commit comes before the other two.
